### PR TITLE
Add ERC: Cross-chain Storage Router Protocol

### PR DESCRIPTION
**Pull request by sshmatrix**
The following proposal is a superior version of [EIP-5559: Cross Chain Write Deferral Protocol](https://ethereum-magicians.org/t/eip-5559-cross-chain-write-deferral-protocol/10576), aka CCIP-Write, replacing Ethereum L1 storage with L2 chains and cryptographically secure databases with an aim to cut gas costs and further privacy while retaining the secure aspects of on-chain storage. Methods in this document specifically target security and cost-effectiveness of write deferrals in context of databases. The cross-chain data written with these methods can be retrieved by generic EIP-3668-compliant contracts completing the cross-chain data life cycle.

### CODE REVIEW

File: ERCS/erc-____.md [hunk 1]
```diff
@@ -0,0 +1,492 @@
+---
+eip: ____
```
Comment (abcoathup) [inline ERCS/erc-____.md:2]:
```suggestion
eip: 7700
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500.

Please also update the filename.

File: ERCS/erc-____.md [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+title: "Cross-chain Storage Router Protocol"
```
Comment (abcoathup) [inline ERCS/erc-____.md:3]:
```suggestion
title: Cross-chain Storage Router Protocol
```
Quotes are not required

File: ERCS/erc-____.md (hunk 1 above)
Comment (sshmatrix) [reply ERCS/erc-____.md:2]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/7af50f32660600c2321aa5ba29af24322bd8c304

File: ERCS/erc-____.md (hunk 2 above)
Comment (sshmatrix) [reply ERCS/erc-____.md:3]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/7af50f32660600c2321aa5ba29af24322bd8c304

File: ERCS/erc-7700.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+simultaneously in the background without ever prompting the user. Signature(s) `dataSig` accompanying the off-chain calldata must implement the following format in their message payloads:  
+
+```text
+Requesting Signature To Update Off-Chain Data\n\nOrigin: ${username}\nData Type: ${dataType}\nData Value: ${dataValue}
+```
+
+where `dataType` parameters are protocol-specific and formatted as object keys delimited by `/`. For instance, if the off-chain data is nested in keys as `a > b > c > field > key`, then the equivalent `dataType` is `a/b/c/field/key`. For example, in order to update off-chain ENS record `text > avatar` and `address > 60`, `dataType` must be formatted as `text/avatar` and `address/60` respectively.
+ 
+#### 3. Approve Data Signer
+To further save on gas costs, we do not store the `dataSigner` on L1. Instead, the clients must
```
Comment (SamWilsn) [inline ERCS/erc-7700.md:313]: > To further save on gas costs, we do not store the `dataSigner` on L1.

This content belongs in the Rationale section. Keep the Specification section as focused as possible to make it easier for implementers to follow.

File: ERCS/erc-7700.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+
+## Security Considerations
+1. Clients must purge the derived signer private keys from local storage immediately after signing the off-chain data.
+
+2. Signature message payload and the resulting deterministic signature `sigKeygen` must be treated as a secret by the clients and immediately purged from local storage after usage in the `keygen()` function.
+
+3. Clients must immediately purge the `password` from local storage after usage in the `keygen()` function.
+
+## Copyright
+Copyright and related rights waived via [`CC0`](../LICENSE.md).
```
Comment (SamWilsn) [inline ERCS/erc-7700.md:549]:
```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

Use the exact text from the template.

File: ERCS/erc-7700.md [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+discussions-to: https://ethereum-magicians.org/t/erc-7700-cross-chain-storage-router-protocol/19853
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-04-30
+requires: 155
+---
+
+## Abstract
+The following standard provides a superior mechanism to [EIP-5559](./eip-5559), aka CCIP-Write, by which smart contracts can route storage to external providers. In particular, protocols can reduce the gas fees associated with storing data on mainnet by routing the handling of storage operations to another system or network. These storage routers act as an extension to the core L1 contract. Methods in this document specifically target security and cost-effectiveness of storage routing to three router types: L1, L2 and databases. The cross-chain data written with these methods can be retrieved by generic [EIP-3668](./eip-3668)-compliant contracts, thus completing the cross-chain data life cycle. This document, nicknamed CCIP-Store, alongside [EIP-3668](./eip-3668), is a meaningful step toward a secure infrastructure for cross-chain storage routers and data retrievals.
```
[the remark this answers is not in the document]
Comment (sshmatrix) [reply ERCS/erc-7700.md]: Removed reference to EIP-5559 in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

File: ERCS/erc-7700.md (hunk 4 above)
Comment (sshmatrix) [reply ERCS/erc-7700.md]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

File: ERCS/erc-7700.md (hunk 3 above)
Comment (sshmatrix) [reply ERCS/erc-7700.md]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

File: ERCS/erc-7700.md [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+  let hashKey = hkdf(sha256, inputKey, salt, username, 42);
+  // Calculate and return secp256k1 keypair
+  return secp256k1(hashKey) // Calculate secp256k1 keypair from hash key
+}
+```
+
+This `keygen()` function requires three variables: `username`, `password` and `sigKeygen`. Their definitions are given below.
+
+##### `username`
+CAIP-10 identifier `username` is auto-derived from the connected wallet's checksummed address `wallet` and `chainId` using [EIP-155](./eip-155).
```
[the remark this answers is not in the document]
Comment (sshmatrix) [reply ERCS/erc-7700.md]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

File: ERCS/erc-7700.md [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+    bytes32 node,
+    bytes32 key,
+    bytes32 value
+) external {
+    // Some code storing data mapped by node & msg.sender
+    ...
+}
+```
+
+```
```
[the remark this answers is not in the document]
Comment (sshmatrix) [reply ERCS/erc-7700.md]: SVG figures added in assets in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

File: ERCS/erc-7700.md [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+
+In this context, a specification which allows storage routing to external routers will facilitate creation of services that are agnostic to the underlying storage solution. This in turn enables new applications to operate without knowledge of the underlying routers. This 'CCIP-Store' proposal outlines precisely this part of the process, i.e. how the bespoke storage routing can be made by smart contracts to L2s and databases. 
+
+## Specification
+### Overview
+The following specification revolves around the structure and description of a cross-chain storage router tasked with the responsibility of writing to an L2 or database storage. This draft introduces `StorageRoutedToL2()` and `StorageRoutedToDatabase()` storage routers, and proposes that new `StorageRoutedTo__()` reverts be allowed through new EIPs that sufficiently detail their interfaces and designs. Some foreseen examples of new storage routers include `StorageRoutedToSolana()` for Solana, `StorageRoutedToFilecoin()` for Filecoin, `StorageRoutedToIPFS()` for IPFS, `StorageRoutedToIPNS()` for IPNS, `StorageRoutedToArweave()` for Arweave, `StorageRoutedToArNS()` for ArNS, `StorageRoutedToSwarm()` for Swarm etc.
```
[the remark this answers is not in the document]
Comment (sshmatrix) [reply ERCS/erc-7700.md]: Fixed in https://github.com/ethereum/ERCs/pull/404/commits/363fc77d6fbba3919a18aaab2ba757e9af9fae42

**Comment by sshmatrix** [comment]
Hi @SamWilsn,

SVGs are included and your comments have been merged. Thanks for the review! Please let us know if you spot any more corrections.

File: ERCS/erc-7700.md [hunk 9]
```diff
@@ -0,0 +1,507 @@
+---
+eip: 7700
+title: Cross-chain Storage Router Protocol
+description: The cross-chain storage router protocol provides a mechanism to replace L1 storage with L2 and databases through off-chain routers
```
Comment (SamWilsn) [inline ERCS/erc-7700.md:4]:
```suggestion
description: Provides a mechanism to replace L1 storage with L2 and databases through off-chain routers
```

Can trim this down a bit, I think.

**Review by SamWilsn — APPROVED** [review]
Apply (or don't) my previous suggestion at your leisure.

**Comment by sshmatrix** [comment]
> Apply (or don't) my previous suggestion at your leisure.

Noting this down; will add to a future PR (when/if one comes along)

**Pull request merged**